C++/Alternate Encoder/src/main/cpp/Robot.cpp

```diff
@@ -7,121 +7,138 @@
+        frc::SmartDashboard::PutNumber("Min Output", kMinOutput);
+        frc::SmartDashboard::PutNumber("Set Rotations", 0);
+
+        /**
+         * The ResetMode::kResetSafeParameters constant can be used to reset
+         * the configuration parameters in the SPARK MAX to their factory
+         * default state. If PersistMode::kNoPersistParameters is passed,
+         * these parameters will not persist between power cycles.
+         */
+        m_motor.Configure(maxConfig,
+            rev::spark::SparkMax::ResetMode::kResetSafeParameters,
+            rev::spark::SparkMax::PersistMode::kNoPersistParameters);
     }
 
-    /**
-     * PIDController objects are commanded to a set point using the 
-     * SetReference() method.
-     * 
-     * The first parameter is the value of the set point, whose units vary
-     * depending on the control type set in the second parameter.
-     * 
-     * The second parameter is the control type can be set to one of four 
-     * parameters:
-     *  rev::CANSparkMax::ControlType::kDutyCycle
-     *  rev::CANSparkMax::ControlType::kPosition
-     *  rev::CANSparkMax::ControlType::kVelocity
-     *  rev::CANSparkMax::ControlType::kVoltage
-     */
-    m_pidController.SetReference(rotations, rev::CANSparkMax::ControlType::kPosition);
-    
-    frc::SmartDashboard::PutNumber("SetPoint", rotations);
-    frc::SmartDashboard::PutNumber("ProcessVariable", m_alternateEncoder.GetPosition());
-    
-  }
+    void TeleopPeriodic() {
+        // read PID coefficients from SmartDashboard
+        double p = frc::SmartDashboard::GetNumber("P Gain", 0);
+        double i = frc::SmartDashboard::GetNumber("I Gain", 0);
+        double d = frc::SmartDashboard::GetNumber("D Gain", 0);
+        double iz = frc::SmartDashboard::GetNumber("I Zone", 0);
+        double ff = frc::SmartDashboard::GetNumber("Feed Forward", 0);
+        double max = frc::SmartDashboard::GetNumber("Max Output", 0);
+        double min = frc::SmartDashboard::GetNumber("Min Output", 0);
+        double rotations = frc::SmartDashboard::GetNumber("Set Rotations", 0);
+
+        // if PID coefficients on SmartDashboard have changed, write new values to controller
+        rev::spark::ClosedLoopConfig pidConfig;
+        if ((p != kP)) { pidConfig.P(p); kP = p; }
+        if ((i != kI)) { pidConfig.I(i); kI = i; }
+        if ((d != kD)) { pidConfig.D(d); kD = d; }
+        if ((iz != kIz)) { pidConfig.IZone(iz); kIz = iz; }
+        if ((ff != kFF)) { pidConfig.VelocityFF(ff); kFF = ff; }
+        if ((max != kMaxOutput) || (min != kMinOutput)) {
+            pidConfig.OutputRange(min, max);
+            kMinOutput = min; kMaxOutput = max;
+        }
+
+        rev::spark::SparkMaxConfig maxConfig;
+        maxConfig.Apply(pidConfig);
+        m_motor.Configure(maxConfig,
+            rev::spark::SparkMax::ResetMode::kNoResetSafeParameters,
+            rev::spark::SparkMax::PersistMode::kNoPersistParameters);
+
```

Comment:
If none of the parameters change, does this still send out the restore defaults? Probably a @jfabellera question. Not sure if we need to check if any of the parameters changed before calling the configure() method.

Reply:
It has `kNoResetSafeParameters` and `kNoPersistParameters` so it doesn't send it at all. The config object is getting reinitialized every loop iteration, and values are only getting set if they are changed. If it gets to the `.Configure()` without setting any values, essentially no CAN traffic is generated (and thus not getting continuously blocked).

This technically works, but we might wanna make it clearer or moving to a button that just applies all the parameters instead of doing this if check thing. Idk, it's not really important to the example, but it's still in the code

Reply:
ah, thought it said kResetSafeParameters :S